drv/stm32h7-spi-rot-server/src/main.rs

```diff
@@ -0,0 +1,274 @@
+    // But, hacking in one interrupt as an example should be ok to start things
+    // off.
+
+    sys_irq_control(self.interrupt, true);
+    // And wait for it to arrive.
+    let _rm =
+        sys_recv_closed(&mut [], self.interrupt, TaskId::KERNEL)
+            .unwrap_lite();
+    */
+
+    // Read just the header.
+    // Keep CSn asserted over the two reads.
+    server
+        .spi
+        .lock(CsState::Asserted)
+        .map_err(|_| MsgError::SpiServerError)?;
+    ringbuf_entry!(Trace::Line);
+    if let Err(spi_error) =
+        server.spi.read(&mut server.message[0..SPI_HEADER_SIZE])
+    {
+        ringbuf_entry!(Trace::SpiError(spi_error));
+        server.spi.release().unwrap_lite();
+        return Err(MsgError::SpiServerError); // XXX spi_error value?
+    }
+
+    ringbuf_entry!(Trace::Data(
+        server.message[0],
+        server.message[1],
+        server.message[2],
+        server.message[3]
+    ));
+
+    let msg = Msg::parse(&mut server.message[..]).unwrap_lite();
+    if !msg.is_supported_version() {
+        ringbuf_entry!(Trace::BadProtocol(server.message[0]));
+        server.spi.release().unwrap_lite();
+        return Err(MsgError::UnsupportedProtocol);
+    }
+    let rlen = msg.payload_len();
+    ringbuf_entry!(Trace::RspLen(rlen));
+    if rlen > server.message.len() - SPI_HEADER_SIZE {
+        ringbuf_entry!(Trace::BadMessageLength(rlen));
+        server.spi.release().unwrap_lite();
+        return Err(MsgError::BadTransferSize);
+    }
+    if let Err(spi_error) = server
+        .spi
+        .read(&mut server.message[SPI_HEADER_SIZE..SPI_HEADER_SIZE + rlen])
+    {
+        ringbuf_entry!(Trace::SpiError(spi_error));
+        server.spi.release().unwrap_lite();
+        return Err(MsgError::SpiServerError);
+    }
+    server.spi.release().unwrap_lite();
+
+    let msg = Msg::parse(&mut server.message[0..rlen + SPI_HEADER_SIZE])
+        .unwrap_lite();
+    match msg.payload_get() {
+        Err(err) => Err(err),
+        Ok(buf) => Ok(buf.len()),
+    }
+}
+
+struct ServerImpl {
+    sys: sys_api::Sys,
+    spi: drv_spi_api::SpiDevice,
+    pub message: [u8; SPI_RSP_BUF_SIZE],
+}
+
+impl idl::InOrderSpiMsgImpl for ServerImpl {
+    /// A client sends a message for SPDM processing.
+    fn send_recv(
+        &mut self,
+        _: &RecvMessage,
+        msgtype: drv_spi_msg::MsgType,
+        source: Leased<R, [u8]>,
+        sink: Leased<W, [u8]>,
+    ) -> Result<[u32; 2], RequestError<MsgError>> {
+        ringbuf_entry!(Trace::Line);
+        let mut msg =
+            drv_spi_msg::Msg::parse(&mut self.message[..]).unwrap_lite();
+
+        msg.set_version();
+        msg.set_len(source.len());
+        msg.set_msgtype(msgtype);
+        ringbuf_entry!(Trace::SendRecv(
+            msgtype,
+            source.len(),
+            msg.payload_buf().len()
+        ));
+        if source.len() > msg.payload_buf().len() {
+            ringbuf_entry!(Trace::Line);
```

Comment:
What's the purpose of this tracing? Shouldn't this return an error?

Reply:
Errors now handled.